src/1-theory/10-async.adoc

```diff
@@ -0,0 +1,407 @@
+
+- https://pipedream.com/[PipeDream] - Formerly known as RequestBin, create an endpoint and register that as the callback URL, and see what traffic is coming in. Then maybe copy that to use in a test suite
+- https://ngrok.com/[ngrok] - Create a tunnel to your locally running application and register your local app as the callback URL
+
+That ngrok example might have made you notice an issue with web hooks: you need
+some sort of DNS or known IP address to fire the payload at. If the client is
+written in PHP, Java, Node, Ruby, etc. and running behind a web server then
+yeah, it will probably have one of those, but if the client is a JavaScript
+application running in the users browser then that isn't going to work.
+
+For that we need another approach, and that is... 🥁
+
+=== WebSocket
+
+WebSocket is a whole other protocol to HTTP, that operates over TCP directly.
+Instead of a client asking the server over and over again, or making a blocking
+connection to the web server, the client establishes a connection to the server,
+and then the two parties can send messages back and forth in real-time, and the
+server doesn't need to know the IP address of the client because the client
+initiated the connection. Cool!
+
+[quote,Wikipedia,https://en.wikipedia.org/wiki/WebSocket]
+----
+Both protocols are located at layer 7 in the OSI model and depend on TCP at layer 4. Although
+they are different, RFC 6455 states that WebSocket "is designed to work over
+HTTP ports 80 and 443 as well as to support HTTP proxies and intermediaries"
+thus making it compatible with the HTTP protocol. To achieve compatibility, the
+WebSocket handshake uses the HTTP Upgrade header to change from the HTTP
+protocol to the WebSocket protocol.
+----
+
+Basically whatever client and server side code you are using will do the same
+thing: the server provides a WebSocket URL to the client, and they connect, and
+messages can then go either way. Just like HTTP has `http://` and `https://`,
+there is `ws://` and `wss://`.
+
+[source,javascript]
+____
+// Create WebSocket connection
+const socket = new WebSocket('ws://localhost:8080/cakes/abc123');
+
+// Connection opened
+socket.addEventListener('open', function (event) {
+    socket.send('Hello Server!');
+});
+
+// Listen for messages
+socket.addEventListener('message', function (event) {
+    console.log('update RE that cake ', event.data);
+});
+____
+
+The fact that HTTP and WS are quite similar in many ways, and that they are
+compatible in general, make it really easy to add WebSocket's into your HTTP API
+by using a web server that has it built in. MDN maintain
+https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API#Tools[a list of
+WebSocket tools] for folks interested in giving it a try, including web servers
+that support them.
+
+You can punt figuring out how to run a WebSocket server entirely, but using
+hosted solutions like https://pusher.com/[Pusher] and
+https://www.pubnub.com/websockets/[PubNub].
```

Comment:
forgot to finish the sentence... "but using hosted solutions like ... " - is what?

Reply:
"by using..." 🤦‍♂️